wrapanapi/systems/rhevm.py

```diff
@@ -1148,31 +1148,94 @@ def _get_attached_storage_domain_service(self, datacenter_id, storage_domain_id)
+    def change_storage_domain_state(self, state, storage_domain_name, timeout=300):
+        """Activate/deactivate storage domain
+        Cannot directly set state to things like 'locked', 'detaching', etc
+
+        Notes:
+            Any state passed that is not 'active' is taken to be a 'deactivate' action
+            Thus any state other than 'active' will result in storage domain being in maintenance
+            Does not wait for steady state before deactivating, simply looks for non-active state
+
+        Args:
+            state (str): valid ovirt.types.StorageDomainStatus enum value
+            storage_domain_name (str): name of the storage domain to modify state on
+            timeout (int): number of seconds to wait for state change
+
+        Returns:
+            None: domain already on given state
+            True: domain changed to given state
+
+        Raises:
+            ValueError: when an invalid StorageDomainStatus enum is passed, or storage domain name
+            TimedOutError: when the desired state is not reached in the timeout
+        """
+        desired_state = getattr(types.StorageDomainStatus, state.upper(), None)
+        active = types.StorageDomainStatus.ACTIVE
+        if desired_state is None:
+            raise ValueError('Invalid state [{}] passed for setting storage domain, '
+                             'value values are {}'.format(state, list(types.StorageDomainStatus)))
+        for datacenter in self._data_centers_service.list():
+            for domain in self.api.follow_link(datacenter.storage_domains):
                 if domain.name == storage_domain_name:
-                    asds = self._get_attached_storage_domain_service(dc.id, domain.id)
-                    if state == "maintenance" and domain.status.value == "active":
-                        asds.deactivate()
-                    elif state == "active" and domain.status.value != "active":
-                        asds.activate()
-                    wait_for(lambda: domain.status.value == state, delay=5, num_sec=240)
+                    attached_service = self._get_attached_storage_domain_service(datacenter.id,
+                                                                                 domain.id)
+                    domain_status = self.api.follow_link(domain).status
+                    if domain_status == desired_state:
+                        return None  # already on the state we wanted
+                    elif desired_state != active:
+                        attached_service.deactivate()
+                        expected_state = types.StorageDomainStatus.MAINTENANCE
+                    else:
+                        attached_service.activate()
+                        expected_state = active
+                    wait_for(
+                        lambda: self.api.follow_link(domain).status == expected_state,
+                        delay=5,
+                        num_sec=timeout
```

Comment:
Will there be error message that explains what happened. If request times out, need to go and check the storage domain status in the RHV UI first before making another request.

Reply:
In the case where the domain does not reach the expected state, this will raise a TimedOutError.  There's no message otherwise, nor was there before.  We're just checking the state of the domain, not tracking an event for the state change that includes some message.  Discerning the reason, if any, for the domain not changing state, is not something I want to address in this PR, or that wrapanapi should necessary do.

This behavior was completely broken before, the domain status property was not dynamically updating - we were referencing a static value in memory.

There's certainly issues with the previous implementation, and this update, where the domain could be in a state not valid for the activate/deactivate call when it is made.  This is not a 'safe' setting of the domain to an explicit state (like with power control methods). I wrote a docblock so the actions taken are clear - there's not steady-state observation or protection, we're just calling the ovirt activate/deactivate depending on if you passed 'active' or anything else.